python/cuml/test/test_kernel_ridge.py

```diff
@@ -0,0 +1,274 @@
+                     wrong_parameter_name=1.0)
+
+    # incorrect function type
+    def non_numba_kernel(x, y):
+        return x.dot(y)
+
+    with pytest.raises(
+        TypeError, match="Kernel function should be a numba device function."
+    ):
+        pairwise_kernels(X, metric=non_numba_kernel)
+
+    # correct function type
+    @cuda.jit(device=True)
+    def numba_kernel(x, y,
+                     special_argument=3.0):
+        return 1 + 2
+
+    pairwise_kernels(X, metric=numba_kernel)
+    pairwise_kernels(X, metric=numba_kernel, special_argument=1.0)
+
+    # malformed function
+    @cuda.jit(device=True)
+    def bad_numba_kernel(x):
+        return 1 + 2
+
+    with pytest.raises(
+        ValueError, match="Expected at least two arguments to kernel function."
+    ):
+        pairwise_kernels(X, metric=bad_numba_kernel)
+
+    # malformed function 2 - No default value
+    @cuda.jit(device=True)
+    def bad_numba_kernel2(x, y, z):
+        return 1 + 2
+
+    with pytest.raises(
+        ValueError, match="Extra kernel parameters "
+        "must be passed as keyword arguments."
+    ):
+        pairwise_kernels(X, metric=bad_numba_kernel2)
+
+    # Precomputed
+    assert np.allclose(X, pairwise_kernels(X, metric="precomputed"))
+
+
+@cuda.jit(device=True)
+def custom_kernel(x, y, custom_arg=5.0):
+    sum = 0.0
+    for i in range(len(x)):
+        sum += (x[i] - y[i]) ** 2
+    return math.exp(-custom_arg * sum) + 0.1
+
+
+test_kernels = sorted(
+    PAIRWISE_KERNEL_FUNCTIONS.keys()) + [custom_kernel]
+
+
+@st.composite
+def kernel_arg_strategy(draw):
+    kernel = draw(st.sampled_from(test_kernels))
+    kernel_func = (
+        PAIRWISE_KERNEL_FUNCTIONS[kernel] if isinstance(
+            kernel, str) else kernel
+    )
+    # Inspect the function and generate some arguments
+    all_func_kwargs = list(
+        inspect.signature(kernel_func.py_func).parameters.values())[
+        2:
+    ]
+    param = {}
+    for arg in all_func_kwargs:
+        # 50% chance we generate this parameter or leave it as default
+        if draw(st.booleans()):
+            continue
+        if isinstance(arg.default, float) or arg.default is None:
+            param[arg.name] = draw(st.floats(0.0, 5.0))
+        if isinstance(arg.default, int):
+            param[arg.name] = draw(st.integers(0, 5))
+
+    return (kernel, param)
+
+
+@st.composite
+def array_strategy(draw):
+    X_m = draw(st.integers(1, 20))
+    X_n = draw(st.integers(1, 10))
+    dtype = draw(st.sampled_from([np.float64, np.float32]))
+    X = draw(arrays(dtype=dtype, shape=(X_m, X_n),
+             elements=st.floats(0, 5, width=32),))
+    if draw(st.booleans()):
+        Y_m = draw(st.integers(1, 20))
+        Y = draw(
+            arrays(dtype=dtype, shape=(Y_m, X_n),
+                   elements=st.floats(0, 5, width=32),)
+        )
+    else:
+        Y = None
+    return (X, Y)
+
+
+@given(kernel_arg_strategy(), array_strategy())
```

Comment:
Love the use of hypothesis here. I'm hoping we will start using it more in cuml.